### ISSUE TITLE
[BUG] nn.Module.train() and eval() have NO effect on TransformerLayer's dropout

### ISSUE THREAD

**Issue by zhouyizhuang-megvii** [microsoft/DeepSpeed#1781]
**Describe the bug**
When performing model evaluation during training, using `model.eval()` doesn't bring any change (to dropout) as expected.

`DeepSpeedTransformerLayer`'s forward dropout behaviour is determined by `self.config.training` (specified during construction) rather than `self.training`.

This may hurt accuracy when users construct a model for training and then call `model.eval()` to disable dropout for evaluation.

**To Reproduce**
N/A

**Expected behavior**
After `model.eval()` and `model.train()`, the dropout in `TransformerLayer`s should be disabled and enabled.

A typical fix should be like
```python
class DeepSpeedTransformerLayer(nn.Module):
    ...

    def train(self, mode: bool = True):
        self.config.training = mode
        return super().train(mode=mode)
```

**ds_report output**
N/A

**Screenshots**
N/A

**System info (please complete the following information):**
N/A

**Launcher context**
N/A

**Docker context**
N/A

**Additional context**
N/A

**Comment by tnq177** [comment]
is this expected behavior @tjruwase ?

**Comment by tnq177** [comment]
I too am having this problem. My code snippet is something like this:

```
model = from_pretrained(...)
engine = deepspeed.initialize(
    model=model,
    config=ds_config,
    model_parameters=None,
    optimizer=None
    lr_scheduler=None
)

module = engine.module
module.eval()
loss = 0.
for input in data_iter():
    loss += module(input)
print(loss)
```

I ran the same script multiple times and got different loss value. Seems like `module.eval()` has no impact on `self.training` in sub-modules?

**Comment by ebarkhordar** [comment]
This looks fixed. `DeepSpeedTransformerLayer.forward` now sets `self.config.training = self.training` on every call (`deepspeed/ops/transformer/transformer.py:410`), so `.eval()` and `.train()` do reach the dropout flag handed to the kernel.

That line came in with #2604 (`e7c14026`, Jan 2023). Same idea as the `train()` override suggested here, just done in `forward`, which also picks up `torch.is_grad_enabled()` on the way through.

I read this rather than ran it, no GPU here to build the transformer kernel. `config.training` is only assigned in two places in the tree though: the config constructor and that line.

Probably closeable.
